File: src/intent/dag.go
package intent

import (
	"fmt"
	"sort"
	"strings"
)

type Node struct {
	Name         string
	Intent       *IntentFile
	Dependencies []*Node
	Dependents   []*Node
	Visited      bool
	InStack      bool
}

type DAG struct {
	nodes map[string]*Node
	roots []*Node
}

func NewDAG() *DAG {
	return &DAG{
		nodes: make(map[string]*Node),
		roots: []*Node{},
	}
}

func (d *DAG) AddNode(name string, intent *IntentFile) *Node {
	if node, exists := d.nodes[name]; exists {
		// Update intent if provided
		if intent != nil {
			node.Intent = intent
		}
		return node
	}
	
	node := &Node{
		Name:         name,
		Intent:       intent,
		Dependencies: []*Node{},
		Dependents:   []*Node{},
	}
	
	d.nodes[name] = node
	d.roots = append(d.roots, node)
	return node
}

func (d *DAG) AddEdge(from, to string) error {
	fromNode := d.AddNode(from, nil)
	toNode := d.AddNode(to, nil)
	
	// Check if edge already exists
	for _, dep := range fromNode.Dependencies {
		if dep.Name == to {
			return nil // Edge already exists
		}
	}
	
	fromNode.Dependencies = append(fromNode.Dependencies, toNode)
	toNode.Dependents = append(toNode.Dependents, fromNode)
	
	// Remove 'from' from roots if it has dependencies
	d.updateRoots()
	
	return nil
}

func (d *DAG) updateRoots() {
	newRoots := []*Node{}
	for _, node := range d.nodes {
		if len(node.Dependencies) == 0 {
			newRoots = append(newRoots, node)
		}
	}
	d.roots = newRoots
}

func (d *DAG) DetectCycles() (bool, []string) {
	// Reset visited flags
	for _, node := range d.nodes {
		node.Visited = false
		node.InStack = false
	}
	
	var cycle []string
	
	var visit func(node *Node, path []string) bool
	visit = func(node *Node, path []string) bool {
		node.Visited = true
		node.InStack = true
		path = append(path, node.Name)
		
		for _, dep := range node.Dependencies {
			if !dep.Visited {
				if visit(dep, path) {
					return true
				}
			} else if dep.InStack {
				// Found a cycle, build the cycle path
				startIdx := 0
				for i, name := range path {
					if name == dep.Name {
						startIdx = i
						break
					}
				}
				cycle = path[startIdx:]
				cycle = append(cycle, dep.Name) // Complete the cycle
				return true
			}
		}
		
		node.InStack = false
		return false
	}
	
	for _, node := range d.nodes {
		if !node.Visited {
			if visit(node, []string{}) {
				return true, cycle
			}
		}
	}
	
	return false, nil
}

func (d *DAG) TopologicalSort() ([]*Node, error) {
	// First check for cycles
	if hasCycle, cycle := d.DetectCycles(); hasCycle {
		return nil, fmt.Errorf("dependency cycle detected: %s", strings.Join(cycle, " -> "))
	}
	
	// Reset visited flags
	for _, node := range d.nodes {
		node.Visited = false
	}
	
	var sorted []*Node
	
	var visit func(node *Node)
	visit = func(node *Node) {
		if node.Visited {
			return
		}
		node.Visited = true
		
		// Visit all dependencies first
		for _, dep := range node.Dependencies {
			visit(dep)
		}
		
		// Add node after its dependencies
		sorted = append(sorted, node)
	}
	
	// Visit all nodes
	for _, node := range d.nodes {
		visit(node)
	}
	
	return sorted, nil
}

func (d *DAG) FindAffectedNodes(nodeName string) []*Node {
	node, exists := d.nodes[nodeName]
	if !exists {
		return []*Node{}
	}
	
	// Reset visited flags
	for _, n := range d.nodes {
		n.Visited = false
	}
	
	affected := make(map[string]*Node)
	
	var visit func(n *Node)
	visit = func(n *Node) {
		if n.Visited {
			return
		}
		n.Visited = true
		affected[n.Name] = n
		
		// Visit all dependents
		for _, dep := range n.Dependents {
			visit(dep)
		}
	}
	
	visit(node)
	
	// Convert to slice and sort
	var result []*Node
	for _, n := range affected {
		result = append(result, n)
	}
	
	sort.Slice(result, func(i, j int) bool {
		return result[i].Name < result[j].Name
	})
	
	return result
}

func (d *DAG) GetNode(name string) (*Node, bool) {
	node, exists := d.nodes[name]
	return node, exists
}

func (d *DAG) GetAllNodes() []*Node {
	var nodes []*Node
	for _, node := range d.nodes {
		nodes = append(nodes, node)
	}
	
	sort.Slice(nodes, func(i, j int) bool {
		return nodes[i].Name < nodes[j].Name
	})
	
	return nodes
}

func (d *DAG) GetRoots() []*Node {
	return d.roots
}

func (d *DAG) BuildFromIntents(intents []*IntentFile) error {
	// First pass: add all nodes
	for _, intent := range intents {
		d.AddNode(intent.Name, intent)
	}
	
	// Second pass: add edges
	for _, intent := range intents {
		for _, dep := range intent.Dependencies {
			if err := d.AddEdge(intent.Name, dep); err != nil {
				return fmt.Errorf("failed to add edge from %s to %s: %w", intent.Name, dep, err)
			}
		}
	}
	
	// Check for missing dependencies
	for _, node := range d.nodes {
		if node.Intent == nil && len(node.Dependents) > 0 {
			dependents := []string{}
			for _, dep := range node.Dependents {
				dependents = append(dependents, dep.Name)
			}
			return fmt.Errorf("missing intent file for %s (required by: %s)", 
				node.Name, strings.Join(dependents, ", "))
		}
	}
	
	return nil
}

func (d *DAG) Visualize() string {
	var builder strings.Builder
	
	builder.WriteString("Dependency Graph:\n")
	builder.WriteString("================\n\n")
	
	// Check for cycles first
	if hasCycle, cycle := d.DetectCycles(); hasCycle {
		builder.WriteString(fmt.Sprintf("Error: dependency cycle detected: %s\n", strings.Join(cycle, " -> ")))
		return builder.String()
	}
	
	// Build a tree visualization
	printed := make(map[string]bool)
	
	var printNode func(node *Node, prefix string, isLast bool)
	printNode = func(node *Node, prefix string, isLast bool) {
		if printed[node.Name] {
			return
		}
		printed[node.Name] = true
		
		// Print current node
		connector := "├── "
		if isLast {
			connector = "└── "
		}
		
		builder.WriteString(fmt.Sprintf("%s%s%s\n", prefix, connector, node.Name))
		
		// Prepare prefix for children
		childPrefix := prefix
		if isLast {
			childPrefix += "    "
		} else {
			childPrefix += "│   "
		}
		
		// Print dependents
		for i, dep := range node.Dependents {
			isLastDep := i == len(node.Dependents)-1
			printNode(dep, childPrefix, isLastDep)
		}
	}
	
	// Start with roots
	roots := d.GetRoots()
	if len(roots) == 0 {
		builder.WriteString("No root nodes found\n")
	} else {
		for i, root := range roots {
			if i > 0 {
				builder.WriteString("\n")
			}
			printNode(root, "", true)
		}
	}
	
	return builder.String()
}

func (d *DAG) VisualizeWithStatus(statusFunc func(string) string) string {
	var builder strings.Builder
	
	builder.WriteString("Dependency Graph with Status:\n")
	builder.WriteString("============================\n\n")
	
	// Get all nodes sorted
	nodes := d.GetAllNodes()
	
	// Build status legend
	builder.WriteString("Legend: [✓] built, [✗] failed, [○] pending, [◐] building, [!] outdated\n\n")
	
	printed := make(map[string]bool)
	
	var printNode func(node *Node, prefix string, isLast bool)
	printNode = func(node *Node, prefix string, isLast bool) {
		if printed[node.Name] {
			return
		}
		printed[node.Name] = true
		
		// Get status
		status := "○" // default pending
		if statusFunc != nil {
			switch statusFunc(node.Name) {
			case "built":
				status = "✓"
			case "failed":
				status = "✗"
			case "building":
				status = "◐"
			case "outdated":
				status = "!"
			}
		}
		
		// Print current node
		connector := "├── "
		if isLast {
			connector = "└── "
		}
		
		builder.WriteString(fmt.Sprintf("%s%s[%s] %s\n", prefix, connector, status, node.Name))
		
		// Prepare prefix for children
		childPrefix := prefix
		if isLast {
			childPrefix += "    "
		} else {
			childPrefix += "│   "
		}
		
		// Print dependents
		for i, dep := range node.Dependents {
			isLastDep := i == len(node.Dependents)-1
			printNode(dep, childPrefix, isLastDep)
		}
	}
	
	// Start with roots
	roots := d.GetRoots()
	if len(roots) == 0 {
		// If no roots, might be a cycle, show all nodes
		for i, node := range nodes {
			if !printed[node.Name] {
				if i > 0 {
					builder.WriteString("\n")
				}
				printNode(node, "", true)
			}
		}
	} else {
		for i, root := range roots {
			if i > 0 {
				builder.WriteString("\n")
			}
			printNode(root, "", true)
		}
	}
	
	return builder.String()
}

File: src/graph/dag.py
"""Directed acyclic graph for target dependency resolution."""

from __future__ import annotations

from dataclasses import dataclass, field

from core.types import Target, TargetStatus


@dataclass
class Node:
    """A node in the dependency graph wrapping a Target."""

    name: str
    target: Target
    dependencies: list[Node] = field(default_factory=list)
    dependents: list[Node] = field(default_factory=list)


STATUS_SYMBOLS: dict[TargetStatus, str] = {
    TargetStatus.BUILT: "+",
    TargetStatus.FAILED: "-",
    TargetStatus.PENDING: " ",
    TargetStatus.BUILDING: "~",
    TargetStatus.OUTDATED: "!",
}


class DAGError(Exception):
    """Raised for graph construction or validation errors."""


class DAG:
    """Directed acyclic graph for dependency resolution among targets."""

    def __init__(self) -> None:
        self.nodes: dict[str, Node] = {}
        self.roots: list[Node] = []

    def add_target(self, target: Target) -> None:
        """Add a target as a node. Raises DAGError on duplicate name."""
        name = target.name
        if name in self.nodes:
            raise DAGError(f"graph: duplicate target '{name}'")
        self.nodes[name] = Node(name=name, target=target)

    def resolve(self) -> None:
        """Resolve dependency edges from each target's intent.depends_on.

        Expands glob patterns (e.g., "core/*") before resolving edges.
        After resolution, updates self.roots to contain all nodes with no
        dependencies.

        Raises DAGError for unknown dependencies or self-dependencies.
        """
        from parser.parser import expand_dependency_globs

        known_names = set(self.nodes.keys())

        for node in self.nodes.values():
            raw_deps = node.target.intent.depends_on
            expanded_deps = expand_dependency_globs(raw_deps, known_names)
            for dep_name in expanded_deps:
                if dep_name == node.name:
                    raise DAGError(
                        f"graph: target '{node.name}' depends on itself"
                    )
                if dep_name not in self.nodes:
                    raise DAGError(
                        f"graph: target '{node.name}' depends on unknown target '{dep_name}'"
                    )
                dep_node = self.nodes[dep_name]
                if dep_node not in node.dependencies:
                    node.dependencies.append(dep_node)
                if node not in dep_node.dependents:
                    dep_node.dependents.append(node)

        self.roots = [n for n in self.nodes.values() if not n.dependencies]

    def detect_cycles(self) -> None:
        """Detect cycles via DFS with a recursion stack.

        Raises DAGError with the cycle path if a cycle is found.
        """
        WHITE, GRAY, BLACK = 0, 1, 2
        color: dict[str, int] = {name: WHITE for name in self.nodes}
        parent: dict[str, str | None] = {name: None for name in self.nodes}

        def dfs(name: str) -> None:
            color[name] = GRAY
            node = self.nodes[name]
            for dep in node.dependencies:
                if color[dep.name] == GRAY:
                    # Reconstruct cycle path
                    cycle = [dep.name, name]
                    cur = name
                    while cur != dep.name:
                        cur = parent[cur]  # type: ignore[assignment]
                        if cur is None:
                            break
                        if cur == dep.name:
                            break
                        cycle.append(cur)
                    cycle.reverse()
                    cycle.append(cycle[0])
                    path = " -> ".join(cycle)
                    raise DAGError(
                        f"graph: dependency cycle detected: {path}"
                    )
                if color[dep.name] == WHITE:
                    parent[dep.name] = name
                    dfs(dep.name)
            color[name] = BLACK

        for name in self.nodes:
            if color[name] == WHITE:
                dfs(name)

    def topological_sort(self) -> list[Target]:
        """Return targets in build order (dependencies before dependents).

        Calls detect_cycles first to ensure the graph is a valid DAG.
        """
        self.detect_cycles()

        in_degree: dict[str, int] = {
            name: len(node.dependencies) for name, node in self.nodes.items()
        }
        queue: list[Node] = [
            self.nodes[name] for name, deg in in_degree.items() if deg == 0
        ]
        # Sort queue for deterministic output
        queue.sort(key=lambda n: n.name)

        result: list[Target] = []
        while queue:
            node = queue.pop(0)
            result.append(node.target)
            # Process dependents sorted by name for determinism
            for dep in sorted(node.dependents, key=lambda n: n.name):
                in_degree[dep.name] -= 1
                if in_degree[dep.name] == 0:
                    queue.append(dep)
            queue.sort(key=lambda n: n.name)

        return result

    def get_affected(self, target_name: str) -> list[Target]:
        """Return all transitive dependents of the given target.

        Useful for invalidation: if target_name changes, all returned targets
        are affected and may need rebuilding.

        Raises DAGError if the target is not found.
        """
        if target_name not in self.nodes:
            raise DAGError(f"graph: unknown target '{target_name}'")

        visited: set[str] = set()
        result: list[Target] = []

        def walk(node: Node) -> None:
            for dep in node.dependents:
                if dep.name not in visited:
                    visited.add(dep.name)
                    result.append(dep.target)
                    walk(dep)

        start = self.nodes[target_name]
        walk(start)
        return result

    def get_dependency_chain(self, target_name: str) -> list[Target]:
        """Return the target and all its transitive dependencies in topological order.

        Raises DAGError if the target is not found.
        """
        if target_name not in self.nodes:
            raise DAGError(f"graph: unknown target '{target_name}'")

        visited: set[str] = set()

        def collect(node: Node) -> None:
            if node.name in visited:
                return
            visited.add(node.name)
            for dep in node.dependencies:
                collect(dep)

        start = self.nodes[target_name]
        collect(start)

        # Build a sub-DAG topological sort over collected nodes
        in_degree: dict[str, int] = {}
        for name in visited:
            node = self.nodes[name]
            count = sum(1 for d in node.dependencies if d.name in visited)
            in_degree[name] = count

        queue = sorted(
            [name for name, deg in in_degree.items() if deg == 0]
        )
        result: list[Target] = []
        while queue:
            name = queue.pop(0)
            node = self.nodes[name]
            result.append(node.target)
            for dep in sorted(node.dependents, key=lambda n: n.name):
                if dep.name in in_degree:
                    in_degree[dep.name] -= 1
                    if in_degree[dep.name] == 0:
                        queue.append(dep.name)
            queue.sort()

        return result

    def visualize(self) -> str:
        """Return a human-readable tree representation of the DAG.

        Status symbols: [+] built, [-] failed, [ ] pending, [~] building, [!] outdated
        """
        if not self.nodes:
            return "(empty graph)"

        lines: list[str] = []
        visited: set[str] = set()

        def render(node: Node, prefix: str, is_last: bool, is_root: bool) -> None:
            sym = STATUS_SYMBOLS.get(node.target.status, " ")
            connector = "" if is_root else ("└── " if is_last else "├── ")
            lines.append(f"{prefix}{connector}[{sym}] {node.name}")

            if node.name in visited:
                # Already rendered children elsewhere; skip to avoid duplication
                return
            visited.add(node.name)

            child_prefix = prefix if is_root else (prefix + ("    " if is_last else "│   "))
            dependents = sorted(node.dependents, key=lambda n: n.name)
            for i, child in enumerate(dependents):
                is_last_child = i == len(dependents) - 1
                render(child, child_prefix, is_last_child, False)

        roots = sorted(self.roots, key=lambda n: n.name) if self.roots else []
        # If roots not populated (resolve not called), fall back to all nodes
        if not roots:
            roots = sorted(self.nodes.values(), key=lambda n: n.name)

        for i, root in enumerate(roots):
            render(root, "", i == len(roots) - 1, True)

        return "\n".join(lines)


def new_dag() -> DAG:
    """Create and return an empty DAG."""
    return DAG()
